Stratis.SmartContracts/UInt128.cs

```diff
@@ -0,0 +1,166 @@
+        }
+
+        public static UInt128 Parse(string str)
+        {
+            return new UInt128(str);
+        }
+
+        public UInt128(ulong b)
+        {
+            this.value = new UIntBase(WIDTH, b);
+        }
+
+        public UInt128(byte[] vch, bool lendian = true)
+        {
+            this.value = new UIntBase(WIDTH, vch, lendian);
+        }
+
+        public static UInt128 operator >>(UInt128 a, int shift)
+        {
+            return new UInt128(a.value.ShiftRight(shift));
+        }
+
+        public static UInt128 operator <<(UInt128 a, int shift)
+        {
+            return new UInt128(a.value.ShiftLeft(shift));
+        }
+
+        public static UInt128 operator -(UInt128 a, UInt128 b)
+        {
+            return new UInt128(a.value.Subtract(b.value.GetValue()));
+        }
+
+        public static UInt128 operator +(UInt128 a, UInt128 b)
+        {
+            return new UInt128(a.value.Add(b.value.GetValue()));
+        }
+
+        public static UInt128 operator *(UInt128 a, UInt128 b)
+        {
+            return new UInt128(a.value.Multiply(b.value.GetValue()));
+        }
+
+        public static UInt128 operator /(UInt128 a, UInt128 b)
+        {
+            return new UInt128(a.value.Divide(b.value.GetValue()));
+        }
+
+        public static UInt128 operator %(UInt128 a, UInt128 b)
+        {
+            return new UInt128(a.value.Mod(b.value.GetValue()));
+        }
+
+        public UInt128(byte[] vch) : this(vch, true)
+        {
+        }
+
+        public static bool operator <(UInt128 a, UInt128 b)
+        {
+            return UIntBase.Comparison(a.value, b.value) < 0;
+        }
+
+        public static bool operator >(UInt128 a, UInt128 b)
+        {
+            return UIntBase.Comparison(a.value, b.value) > 0;
+        }
+
+        public static bool operator <=(UInt128 a, UInt128 b)
+        {
+            return UIntBase.Comparison(a.value, b.value) <= 0;
+        }
+
+        public static bool operator >=(UInt128 a, UInt128 b)
+        {
+            return UIntBase.Comparison(a.value, b.value) >= 0;
+        }
+
+        public static bool operator ==(UInt128 a, UInt128 b)
+        {
+            return UIntBase.Comparison(a.value, b.value) == 0;
+        }
+
+        public static bool operator !=(UInt128 a, UInt128 b)
+        {
+            return !(a == b);
+        }
+
+        public static implicit operator UInt128(ulong value)
+        {
+            return new UInt128(value);
+        }
+
+        public static implicit operator UInt128(long value)
+        {
+            if (value < 0)
+                throw new ArgumentException("Only positive or zero values are allowed.", nameof(value));
+
+            return new UInt128((ulong)value);
+        }
+
+        public static implicit operator UInt128(int value)
+        {
+            if (value < 0)
+                throw new ArgumentException("Only positive or zero values are allowed.", nameof(value));
```

Comment:
Implicity or explicit conversations always throws OverflowException for numeric types and i think we should to follow same behavior.